### PR TITLE
5137 cepconnector error handling

### PR DESCRIPTION
**Pull request by Andyck**

### CODE REVIEW

File: src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java [hunk 1]
```diff
@@ -285,7 +285,20 @@ protected void doInitialSubscriptions() {
             obj.addProperty("subscription", channel);
             body.add(obj);
         }
-        cloudOfThingsRestClient.doPostRequest(body.toString(), notificationPath, CONTENT_TYPE);
+        String responseBody = cloudOfThingsRestClient.doPostRequest(body.toString(), notificationPath, CONTENT_TYPE);
+        JsonArray responseJson = gson.fromJson(responseBody, JsonArray.class);
+        for (JsonElement channelJson : responseJson) {
+            JsonObject channelObject = (JsonObject) channelJson;
+
+            if (!(channelObject.has("successful") && channelObject.get("successful").getAsBoolean())) {
```
Comment (GannaKrupii) [inline src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: Here the condition is not very clear for understanding, maybe better to separate it

Comment (Matthimatiker) [reply src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: Might be easier to read when the outer braces are removed:

    if (!channelObject.has("successful") || !channelObject.get("successful").getAsBoolean())

File: src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java [hunk 2]
```diff
@@ -285,7 +285,20 @@ protected void doInitialSubscriptions() {
...
+                RuntimeException e = new CotSdkException("Subscription failed! " + channelObject.toString());
+                for (SubscriptionListener listener : listeners) {
+                    listener.onError("", e);
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: Is there a better option for the channel name that is passed to the listener? 
E.g.

- null (when it is not available)
- "system"?
- "internal"?
- "error"?

File: src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java [hunk 3]
```diff
@@ -312,8 +325,21 @@ public void run() {
...
+                            JsonElement success = jsonObject.get("successful");
+                            JsonElement advice = jsonObject.get("advice");
+                            if (success != null
+                                    && !success.getAsBoolean()
+                                    && advice != null
+                                    && "handshake".equals(advice.getAsJsonObject().get("reconnect").getAsString())) {
+
+                                Exception e = new CotSdkException("Connection failed! Redo handshake.");
+                                for (SubscriptionListener listener : listeners) {
+                                    listener.onError("", e);
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: Same as above regarding the channel.

File: src/test/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClientTest.java [hunk 4]
```diff
@@ -23,7 +23,7 @@
 public class CloudOfThingsRestClientTest extends PowerMockTestCase {
 
     @Test(expectedExceptions = CotSdkException.class)
-    public void testDoRequetsWithIdResponseWithException() throws Exception {
+    public void testDoRequestsWithIdResponseWithException() throws Exception {
```
Comment (Matthimatiker) [inline src/test/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClientTest.java]: Nice catch :-)

File: src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java [hunk 5]
```diff
@@ -105,5 +109,94 @@ public String answer(InvocationOnMock invocation) {
         assertNull(connector.getClientId());
     }
 
+    @Test
+    public void testSubscriptionFailed() throws InterruptedException {
+        // return clientId on handshake
+        when(cloudOfThingsRestClient.doPostRequest(eq("[{\"channel\":\"/meta/handshake\",\"version\":\"1.0\",\"minimumVersion\":\"1.0\",\"supportedConnectionTypes\":[\"long-polling\"],\"advice\":{\"timeout\":1000,\"interval\":10}}]"), anyString(), anyString())).
```
Comment (Matthimatiker) [inline src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java]: Is there an easy way to relax the argument check?

What if the order of properties changes or a property is added? That could cause the test to break, but it's most probably not a problem for the tested code.

File: src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java [hunk 6]
```diff
@@ -105,5 +109,94 @@ public String answer(InvocationOnMock invocation) {
...
+        connector.subscribe("/123");
+        connector.addListener(new SubscriptionListener() {
+            @Override
+            public void onNotification(String channel, Notification notification) {
+                notedOperations.add(notification.getData().toString());
+            }
+
+            @Override
+            public void onError(String channel, Throwable error) {
+                notedError = true;
```
Comment (Matthimatiker) [inline src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java]: :+1: Nice check!

File: src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java (hunk 1 above)
Comment (Andyck) [reply src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: already done

File: src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java (hunk 5 above)
Comment (Andyck) [reply src/test/java/com/telekom/m2m/cot/restsdk/realtime/CepConnectorTest.java]: good suggestion!
easier matcher was used!

File: src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java (hunk 2 above)
Comment (Andyck) [reply src/main/java/com/telekom/m2m/cot/restsdk/realtime/CepConnector.java]: done: "null" is a good option.

**Pull request merged**